example/bank/cmd/api/account.go

```diff
@@ -0,0 +1,240 @@
+
+func newAccountHandler(db *sql.DB, repo *domain.BankAccountRepository, logger *zap.Logger) *accountHandler {
+	return &accountHandler{db, repo, logger}
+}
+
+type accountHandler struct {
+	db              *sql.DB
+	bankAccountRepo *domain.BankAccountRepository
+	logger          *zap.Logger
+}
+
+func (a *accountHandler) HTTPHandler() *chi.Mux {
+	r := chi.NewRouter()
+	r.Use(middleware.NoCache)
+
+	r.Post("/", a.CreateAccount)
+
+	r.Route("/{account_id}", func(r chi.Router) {
+		r.Use(a.accountIDFromURL)
+
+		r.Get("/", a.ViewAccount)
+		r.Get("/average", a.AccountAverages)
+		r.Post("/deposit", a.AccountDeposit)
+		r.Post("/withdraw", a.AccountWithdraw)
+	})
+
+	return r
+}
+
+func (a *accountHandler) CreateAccount(w http.ResponseWriter, r *http.Request) {
+	bankAccount, err := domain.OpenBankAccount()
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		a.logger.With(zap.Error(err)).Error("failed to open a bank account")
+		return
+	}
+
+	if err := a.bankAccountRepo.Save(r.Context(), bankAccount); err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		a.logger.With(zap.Error(err)).Error("failed to save a opened bank account")
+		return
+	}
+
+	w.Header().Add("Location", fmt.Sprintf("/account/%s", bankAccount.AggregateID()))
+	w.WriteHeader(http.StatusCreated)
+}
+
+func (a *accountHandler) ViewAccount(w http.ResponseWriter, r *http.Request) {
+	bankAccount := a.loadAccount(w, r)
+	if bankAccount == nil {
+		return
+	}
+	body := []byte(
+		fmt.Sprintf(`{"version":%d}`, bankAccount.AggregateVersion()),
+	)
+
+	w.Header().Add("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+	if _, err := w.Write(body); err != nil {
+		a.logger.With(zap.Error(err)).Error("failed to write response")
+	}
+}
+
+func (a *accountHandler) AccountAverages(w http.ResponseWriter, r *http.Request) {
+	var (
+		debit  uint
+		credit uint
+	)
+	accountID := r.Context().Value(contextKeyBankAccountID).(aggregate.ID)
+	err := a.db.QueryRowContext(r.Context(),
+		"SELECT debit, credit FROM account_averages WHERE accountid=$1",
+		accountID,
+	).Scan(&debit, &credit)
+	if err != nil {
+		if err != sql.ErrNoRows {
+			w.WriteHeader(http.StatusNotFound)
+			a.logger.With(zap.Error(err)).Debug("unknown accountID provided")
+			return
+		}
+
+		w.WriteHeader(http.StatusInternalServerError)
+		a.logger.Error("failed to query account averages")
+		return
+	}
+	body := []byte(
+		fmt.Sprintf(`{"average_deposit":%d,"average_credit":%d}`, debit, credit),
+	)
+
+	w.Header().Add("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+	if _, err = w.Write(body); err != nil {
+		a.logger.With(zap.Error(err)).Error("failed to write response")
+	}
+}
+
+func (a *accountHandler) AccountDeposit(w http.ResponseWriter, r *http.Request) {
+	amount, err := a.loadAmountFromPost(w, r)
+	if err != nil {
+		return
+	}
+
+	bankAccount := a.loadAccount(w, r)
+	if bankAccount == nil {
+		return
```

Comment:
No error message to display?

Reply:
The error is done by `loadAccount` I didn't want to duplicate but it may indeed be a little confusing.... Let me see if I can improve this.